File: code/python/data_structures_article-07.py
"""Binary Tree Mental Method (Construction)"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().buildTree(*args, **kwargs)


class _Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        valToIndex = {}
        for i in range(len(inorder)):
            valToIndex[inorder[i]] = i
        return self.build(preorder, 0, len(preorder) - 1,
                          inorder, 0, len(inorder) - 1, valToIndex)
    def build(self, preorder, preStart, preEnd, inorder, inStart, inEnd, valToIndex):
        if preStart > preEnd:
            return None
        rootVal = preorder[preStart]
        index = valToIndex[rootVal]
        leftSize = index - inStart
        root = TreeNode(rootVal)
        root.left = self.build(preorder, preStart + 1, preStart + leftSize,
                               inorder, inStart, index - 1, valToIndex)
        root.right = self.build(preorder, preStart + leftSize + 1, preEnd,
                                inorder, index + 1, inEnd, valToIndex)
        return root

File: code/javascript/data_structures_article-07.js
/**
 * Binary Tree Mental Method (Construction)
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return buildTree(...args);
}

var buildTree = function(preorder, inorder) {
    var valToIndex = new Map();
    for (var i = 0; i < inorder.length; i++) {
        valToIndex.set(inorder[i], i);
    }
    return build(preorder, 0, preorder.length - 1,
                inorder, 0, inorder.length - 1);
};
var build = function(preorder, preStart, preEnd,
               inorder, inStart, inEnd) {
    if (preStart > preEnd) {
        return null;
    }
    var rootVal = preorder[preStart];
    var index = valToIndex.get(rootVal);
    var leftSize = index - inStart;
    var root = new TreeNode(rootVal);
    root.left = build(preorder, preStart + 1, preStart + leftSize,
            inorder, inStart, index - 1);
    root.right = build(preorder, preStart + leftSize + 1, preEnd,
            inorder, index + 1, inEnd);
    return root;
};
